src/WebJobs.Script/Host/WorkerFunctionMetadataProvider.cs

```diff
@@ -74,12 +77,25 @@ public async Task<FunctionMetadataResult> GetFunctionMetadataAsync(IEnumerable<R
 
                 var channels = _channelManager.GetChannels(_workerRuntime);
 
+                // Start up GRPC channels if they are not already running.
                 if (channels?.Any() != true)
                 {
-                    await _channelManager.InitializeChannelAsync(workerConfigs, _workerRuntime);
+                    if (_scriptHostManager.State is ScriptHostState.Default
+                        || _scriptHostManager.State is ScriptHostState.Starting
+                        || _scriptHostManager.State is ScriptHostState.Initialized)
+                    {
+                        // We don't need to restart if the host hasn't even been created yet.
+                        await _channelManager.InitializeChannelAsync(workerConfigs, _workerRuntime);
```

Comment:
This is definitely better than before, but it still seems like there's a chance for races... can we add some logging so it's clear what we're doing here and below (with the restart)? Just in case we need to debug in the future, it'd be good to know the path we took through this code...